Add wimprates rate generator to nestWIMPSource

lorenzomag opened the pull request:
This PR proposes to utilise the `wimprates` package to compute the differential rates of WIMP-nucleus interaction instead of fetching them from pickle files.

If `modulation==True`, the differential rate (with respect to energy and time, per unit mass) will take into consideration annual modulation patterns, calculated within wimprates itself.

Initial and final timestamps are provided by the user using keyword arguments to the class initialiser method.
The livetime is computed from their difference.

If `modulation==False`, a conservative timestamp is chosen by wimprates to obtain the average annual rate. The differential rate will then be calculated once, and the result repeated into an array.
This option is more time efficient.
A multiprocessing implementation of the former case should be implemented.

**Currently, there is a discrepancy in the output rate by the function wrt the pickled file's energy spectrum by a factor of 10 approximately!**
This needs to be investigated

lorenzomag committed:
Initial draft of Migdal source

lorenzomag committed:
Initial draft of Migdal source

lorenzomag committed:
Initial draft of Migdal source

lorenzomag committed:
Use multiprocessing when modulation on

lorenzomag committed:
Use multiprocessing when modulation on

lorenzomag committed:
Fix normalisation bug

lorenzomag committed:
Fix normalisation bug

lorenzomag committed:
Fix normalisation bug

lorenzomag committed:
Merge branch 'RJ-XLZD_simple' of github.com:lorenzomag/flamedisx into RJ-XLZD_simple

lorenzomag committed:
Revert "Revert "Merge branch 'RJ-XLZD_simple' of github.com:FlamTeam/flamedisx into RJ-XLZD_simple""

This reverts commit 65da548dd24e3396c9707322588f78ce6320eca9.

lorenzomag committed:
Revert "Revert "Merge branch 'RJ-XLZD_simple' of github.com:FlamTeam/flamedisx into RJ-XLZD_simple""

This reverts commit 65da548dd24e3396c9707322588f78ce6320eca9.

lorenzomag committed:
Merge branch 'multip' into RJ-XLZD_simple

lorenzomag committed:
Updating Migdal source to match WIMP source

lorenzomag committed:
Updating Migdal source to match WIMP source

lorenzomag committed:
Merge branch 'migdal' of github.com:lorenzomag/flamedisx into migdal

lorenzomag committed:
Remove duplicated class

lorenzomag committed:
Remove duplicated class

lorenzomag committed:
Update XLZD sources

lorenzomag committed:
Update XLZD sources

lorenzomag committed:
Fix scaling bug

Scaling was only applied when modulation off

lorenzomag committed:
Fix scaling bug

Scaling was only applied when modulation off

robertsjames commented on flamedisx/nest/lxe_sources.py:
```diff
@@ -552,23 +555,262 @@ class nestWIMPSource(nestNRSource):
+        time_bin_width = time_bin_width.value * nu.ns
+        times = (time_bin_edges[:-1] + time_bin_edges[1:]) / 2
+
+        scale = time_bin_width / nu.year  # Convert from [per year] to [per time_bin_width]
```
@lorenzomag I think we need to get rid of this. I tried to explain on Slack (maybe you missed it): whilst in principle keeping this and removing  `/ self.n_time_bins` would solve things at the level of `source.mu_before_efficiencies()`, it will mess up the tensor-based calculation of the differential rate, as the thing that is queried should be the spectrum corresponding to a given event time (such that the sum of the spectrum gives the expected counts within that energy range, after scaling by the exposure). We don't want to break compatibility with the tensor-based calculation

robertsjames commented on flamedisx/nest/lxe_sources.py:
```diff
@@ -552,23 +555,262 @@ class nestWIMPSource(nestNRSource):
+        super().__init__(*args, **kwargs)
+
+    @staticmethod
+    def get_energy_hist(
```
@lorenzomag can we avoid duplicating all this code?

lorenzomag replied:
Hello, yes that is definitely needed. I just wasn't sure where to put the function. Shall I put it in some utils.py file? Or just keep it in there but out of the classes?

lorenzomag replied:
Ok, I refactored the code and made a new module. The name of the module is just the name of the function for now, but we can change it to utils.py if needed, I guess

robertsjames commented on flamedisx/nest/lxe_sources.py:
```diff
@@ -552,23 +555,262 @@ class nestWIMPSource(nestNRSource):
+        time_bin_width = time_bin_width.value * nu.ns
+        times = (time_bin_edges[:-1] + time_bin_edges[1:]) / 2
+
+        scale = time_bin_width / nu.year  # Convert from [per year] to [per time_bin_width]
```
@lorenzomag same comment as before (though ideally we'll remove the duplicated code)

robertsjames commented on flamedisx/nest/lxe_blocks/energy_spectrum.py:
```diff
@@ -449,7 +449,7 @@ def clip_j2000_times(self, ts):
 
     def mu_before_efficiencies(self, **params):
-        return self.energy_hist.n / self.n_time_bins
+        return self.energy_hist.n
```
@lorenzomag see my other comment: we want to put this back in

lorenzomag committed:
Code refactoring and normalisation fix

The energy histogram is not normalised in the time dimension.

Robert's expalation:  I think all you need to do is remove
scale = time_bin_width / nu.year  # Convert from [per year] to [per time_bin_width]
Whilst in principle keeping this and removing  / self.n_time_bins would solve things at the level of source.mu_before_efficiencies(), it will mess up the tensor-based calculation of the differential rate, as the thing that is queried should be the spectrum corresponding to a given event time (such that the sum of the spectrum gives the expected counts within that energy range, after scaling by the exposure)

lorenzomag committed:
Code refactoring and normalisation fix

The energy histogram is not normalised in the time dimension.

Robert's expalation:  I think all you need to do is remove
scale = time_bin_width / nu.year  # Convert from [per year] to [per time_bin_width]
Whilst in principle keeping this and removing  / self.n_time_bins would solve things at the level of source.mu_before_efficiencies(), it will mess up the tensor-based calculation of the differential rate, as the thing that is queried should be the spectrum corresponding to a given event time (such that the sum of the spectrum gives the expected counts within that energy range, after scaling by the exposure)

lorenzomag committed:
Merge remote-tracking branch 'upstream/RJ-XLZD_simple' into RJ-XLZD_simple

lorenzomag committed:
Merge branch 'RJ-XLZD_simple' of github.com:lorenzomag/flamedisx into RJ-XLZD_simple

Ananthu-Ravindran committed:
4 fold

Ananthu-Ravindran committed:
sync with upstream

Ananthu-Ravindran committed:
new lce from owen
